### PR TITLE
WIP: Attempt to run

### PR DESCRIPTION
**Pull request by larsoner**

### CODE REVIEW

File: badbaby/processing/run_mnefun.py [hunk 1]
```diff
@@ -18,19 +18,19 @@
...
 except ImportError:
     handler = None
-df = rd.return_dataframes('mmn')[0]
+df = return_dataframes('mmn')[0]
 exclude = defaults.exclude
-df.drop(df[df.index.isin(pd.Series(exclude))].index, inplace=True)
+df.drop(df[df.index.isin(pd.Series(exclude, dtype=int))].index, inplace=True)
 ecg_chs = np.unique(df['ecg'].tolist())
-work_dir = defaults.datapath
+work_dir = defaults.datadir
```
Comment (larsoner) [inline badbaby/processing/run_mnefun.py:33]: var missing without this change

File: badbaby/processing/run_mnefun.py [hunk 2]
```diff
@@ -18,19 +18,19 @@
...
 try:
     # Use niprov as handler for events, or if it's not installed, ignore
     from niprov.mnefunsupport import handler
 except ImportError:
     handler = None
-df = rd.return_dataframes('mmn')[0]
+df = return_dataframes('mmn')[0]
 exclude = defaults.exclude
-df.drop(df[df.index.isin(pd.Series(exclude))].index, inplace=True)
+df.drop(df[df.index.isin(pd.Series(exclude, dtype=int))].index, inplace=True)
```
Comment (larsoner) [inline badbaby/processing/run_mnefun.py:31]: pandas warning about `object` being the new default

File: badbaby/processing/run_mnefun.py [hunk 3]
```diff
@@ -18,19 +18,19 @@
 import numpy as np
 import pandas as pd
 
-import badbaby.return_dataframes as rd
 from badbaby import defaults
+from badbaby.defaults import return_dataframes
```
Comment (larsoner) [inline badbaby/processing/run_mnefun.py:22]: maybe you had a namespace change?

File: badbaby/defaults.py [hunk 4]
```diff
@@ -9,14 +9,15 @@
...
-datadir = '/media/ktavabi/ALAYA/data/ilabs/badbaby/mismatch'
+datadir = '/mnt/bakraid/larsoner/kam/badbaby_data/'
 run_name = 'mmn'
 epoching = (-0.1, 0.6)
 #TODO refilter 0.01-30.0
 lowpass = 55.0
 highpass = 0.1
 peak_window = (.235, .53)
 oddball_stimuli = ['standard', 'ba', 'wa']
+exclude = []
```
Comment (larsoner) [inline badbaby/defaults.py:20]: this var was missing without this line

File: badbaby/processing/run_mnefun.py [hunk 5]
```diff
@@ -75,12 +75,15 @@
         params.dates = [None] * len(params.subjects)
         params.work_dir = work_dir
         params.subject_indices = np.arange(len(params.subjects))
+        try:
+            params.subject_indices = [params.subjects.index('bad_117b')]
+        except ValueError:
+            continue
```
Comment (larsoner) [inline badbaby/processing/run_mnefun.py:81]: Just for testing, try to process 117b

**Comment by larsoner** [comment]
After `pip install -ve .` in the `badbaby` dir, running this I get:
```
python -u badbaby/processing/run_mnefun.py 
    
Using 1200 Hz as sampling rate and
    MEG0113 as ECG surrogate...
    13  Subjects:  ['101', '102', '103', '104', '105', '203', '205', '207', '208a', '209a', '301a', '921a', '925a']
    
Using 1200 Hz as sampling rate and
    MEG0143 as ECG surrogate...
    1  Subjects:  ['302a']
    
Using 1800 Hz as sampling rate and
    MEG0111 as ECG surrogate...
    3  Subjects:  ['116a', '122a', '212']
    
Using 1800 Hz as sampling rate and
    MEG0113 as ECG surrogate...
    24  Subjects:  ['113', '114', '120b', '122b', '124a', '125a', '127a', '131b', '211a', '213', '214', '215b', '223a', '223b', '224a', '226b', '228a', '229a', '309a', '310a', '310b', '921b', '923b', '925b']
    
Using 1800 Hz as sampling rate and
    MEG0141 as ECG surrogate...
    5  Subjects:  ['108', '118', '224b', '225b', '312b']
    
Using 1800 Hz as sampling rate and
    MEG0143 as ECG surrogate...
    22  Subjects:  ['109', '110', '117a', '119b', '123a', '127b', '128a', '128b', '130a', '131a', '209b', '211b', '215a', '218a', '220a', '225a', '231b', '311a', '311b', '312a', '314a', '316b']
Pulling raw files from acquisition machine. 
  Checking for proper remote filenames for bad_117a...
('', '')
Traceback (most recent call last):
  File "badbaby/processing/run_mnefun.py", line 170, in <module>
    mnefun.do_processing(
  File "/home/larsoner/mne_data/mnefun/mnefun/_mnefun.py", line 526, in do_processing
    outs[ii] = func(p, subjects, run_indices)
  File "/home/larsoner/mne_data/mnefun/mnefun/_fetching.py", line 39, in fetch_raw_files
    stdout_ = run_subprocess(
  File "<decorator-gen-2>", line 24, in run_subprocess
  File "/home/larsoner/python/mne-python/mne/utils/misc.py", line 166, in run_subprocess
    raise subprocess.CalledProcessError(p.returncode, command, output)
subprocess.CalledProcessError: Command '['ssh', '-p', '22', 'kasga.ilabs.uw.edu', 'find /brainstudio/bad_baby -type f -regex .*bad_117a_mmn_raw-?[0-9]*.fif$2>&1 | grep -v "Permission denied"']' returned non-zero exit status 1.
```

File: badbaby/processing/run_mnefun.py (hunk 1 above)
Comment (ktavabi) [reply badbaby/processing/run_mnefun.py:33]: This is where `ACQ` files reside locally?

File: badbaby/processing/run_mnefun.py (hunk 2 above)
Comment (ktavabi) [reply badbaby/processing/run_mnefun.py:31]: This doesn't do anything unless `excludes` is amended in `defaults`

File: badbaby/processing/run_mnefun.py (hunk 1 above)
Comment (larsoner) [reply badbaby/processing/run_mnefun.py:33]: You tell me :) I assumed that it is -- the key is that `datapath` is not a var in `defaults.py` but `datadir` is, hence why I had to change it to get it to run without an `AttributeError`

File: badbaby/processing/score.py [hunk 6]
```diff
@@ -4,13 +4,28 @@
 
 from __future__ import print_function
```
Comment (ktavabi) [inline badbaby/processing/score.py:5]: Is this considered `<3.0` backwards compatiable? Any reason I need it?

File: badbaby/processing/score.py [hunk 7]
```diff
@@ -24,44 +39,52 @@ def score(p, subjects):
...
+                with open(tab_file, 'r') as fid:
+                    header = fid.readline().lstrip('#').strip()
+                    header = json.loads(header.replace("'", '"'))
+                assert header['participant'] == exp_subj
+                t_tab = datetime.datetime.strptime(
+                    header['date'], '%Y-%m-%d %H_%M_%S.%f'
+                ).replace(tzinfo=timezone('US/Pacific'))
+                t_raw = raw.info['meas_date']
+                off_minutes = abs((t_raw - t_tab).total_seconds() / 60.)
+                # tone/1 or IDS/2 or syllable/3
```
Comment (ktavabi) [inline badbaby/processing/score.py:64]: Is this naming form `DACQ` settings?

File: badbaby/processing/score.py [hunk 8]
```diff
@@ -24,44 +39,52 @@ def score(p, subjects):
...
+            for ti, tab_file in enumerate(tab_files):
+                with open(tab_file, 'r') as fid:
+                    header = fid.readline().lstrip('#').strip()
+                    header = json.loads(header.replace("'", '"'))
+                assert header['participant'] == exp_subj
+                t_tab = datetime.datetime.strptime(
+                    header['date'], '%Y-%m-%d %H_%M_%S.%f'
+                ).replace(tzinfo=timezone('US/Pacific'))
+                t_raw = raw.info['meas_date']
+                off_minutes = abs((t_raw - t_tab).total_seconds() / 60.)
```
Comment (ktavabi) [inline badbaby/processing/score.py:63]: Is this the variable `offset` between `expyfun` and `acq`?

File: badbaby/processing/score.py [hunk 9]
```diff
@@ -24,44 +39,52 @@ def score(p, subjects):
...
+                # tone/1 or IDS/2 or syllable/3
+                good[ti] = (off_minutes < 120) and \
+                    header['exp_name'] == 'syllable' and \
+                    header['session'] == '3'
+            assert sum(good) == 1, sum(good)
```
Comment (ktavabi) [inline badbaby/processing/score.py:68]: Can you #comment what 65-67 does?

File: badbaby/processing/score.py [hunk 10]
```diff
@@ -24,44 +39,52 @@ def score(p, subjects):
...
+                assert header['participant'] == exp_subj
+                t_tab = datetime.datetime.strptime(
+                    header['date'], '%Y-%m-%d %H_%M_%S.%f'
+                ).replace(tzinfo=timezone('US/Pacific'))
+                t_raw = raw.info['meas_date']
+                off_minutes = abs((t_raw - t_tab).total_seconds() / 60.)
+                # tone/1 or IDS/2 or syllable/3
+                good[ti] = (off_minutes < 120) and \
+                    header['exp_name'] == 'syllable' and \
+                    header['session'] == '3'
```
Comment (ktavabi) [inline badbaby/processing/score.py:67]: `header['session'] == '3'` is likely to fail in cases where session order was reordered to enhance subject cooperation.

File: badbaby/processing/score.py [hunk 11]
```diff
@@ -24,44 +39,52 @@ def score(p, subjects):
...
+            data = read_tab(tab_files[np.where(good)[0][0]])
 
-def reconstruct_events(p, subjects):
-    """Reconstruct events from expyfun tab file """
-    for subj in subjects:
-        print('  Running subject %s... ' % subj, end='')
+            # Correct the triggers
+            new_nums = np.array(
+                [TRIGGER_MAP[d['trial_id'][0][0]] for d in data], int)
+            exp_times = [d['play'][0][1] for d in data]
```
Comment (ktavabi) [inline badbaby/processing/score.py:75]: Also summarize 71-74 and we can collapse into `docstring` with previous gist.

**Review by ktavabi — CHANGES_REQUESTED** [review]
@larsoner Thanks this looks good! Just the documentation notes and merge when ready.

File: badbaby/processing/run_mnefun.py (hunk 2 above)
Comment (ktavabi) [reply badbaby/processing/run_mnefun.py:32]: 86 this. Once things are cleaned up I'll refactor `return_dataframes`.

File: badbaby/processing/run_mnefun.py [hunk 12]
```diff
@@ -50,138 +72,146 @@
...
+                        if bad_channels[0] == 'None':
+                            with open(prebad_file, 'w') as f:
+                                f.write('')
+                        else:
+                            bads = ['%s' % ch for ch in
+                                    bad_channels[0].split(', ')]
+                            with open(prebad_file, 'w') as output:
+                                for ch_name in bads:
+                                    output.write('%s\n' % ch_name)
+            # Set the niprov handler to deal with events:
```
Comment (ktavabi) [inline badbaby/processing/run_mnefun.py:95]: We can 86 improv

File: badbaby/processing/score.py (hunk 7 above)
Comment (larsoner) [reply badbaby/processing/score.py:64]: It's what I observed in the expyfun headers in the .tab files

File: badbaby/processing/score.py (hunk 6 above)
Comment (larsoner) [reply badbaby/processing/score.py:5]: 2.7 is ancient we can just kill it. I'm using `f` syntax that requires 3.6+ anyway...

File: badbaby/processing/score.py (hunk 8 above)
Comment (larsoner) [reply badbaby/processing/score.py:63]: Yep, will add comment

**Comment by larsoner** [comment]
Requires https://github.com/LABSN/mnefun/pull/316

File: badbaby/processing/score.py (hunk 7 above)
Comment (ktavabi) [reply badbaby/processing/score.py]: `tone`, `IDS`, `syllable` are what I referred to as the primary stimuli for the sessions. I was just curious how they would get into the `.tab` header. Presumably inherited from `DACQ` or more likely the session `expyfun.py`.

Comment (larsoner) [reply badbaby/processing/score.py]: Indeed the first line of `.tab` is a `# {...}` that has these values

**Pull request merged**